src/rgw/rgw_lua.cc

```diff
@@ -0,0 +1,135 @@
+#include <lua.hpp>
+#include "rgw_lua.h"
+#include "rgw_common.h"
+#include "common/dout.h"
+#include <iostream>
+#include <mutex>
+
+#define dout_subsys ceph_subsys_rgw
+
+namespace rgw::lua {
+
+class lua_state_guard {
+  lua_State* l;
+public:
+  lua_state_guard(lua_State* _l) : l(_l) {}
+  ~lua_state_guard() {lua_close(l);}
+  void reset(lua_State* _l=nullptr) {l = _l;}
+};
+
+int RGWLog(lua_State* L) {
+  auto cct = (CephContext*)lua_touserdata(L, lua_upvalueindex(1));
+
+  constexpr auto NUM_RETURN = 0;
+
+  if (!lua_isstring(L, -1)) {
+    ldout(cct, 1) << "Lua ERROR: missing 'message' parameter when calling Log" << dendl;
+    std::cout << "Lua ERROR: missing 'message' parameter when calling Log"  << std::endl;
+    return NUM_RETURN;
+  }
+  const char* message = lua_tostring(L, -1);
+  if (cct) {
+    ldout(cct, 20) << "Lua INFO: " << message << dendl;
+    std::cout << "Lua INFO: " << message << std::endl;
+  }
+  return NUM_RETURN;
+};
+
+// closure that expect req_state as the "up value"
+int S3RequestIndex(lua_State* L) {
+  auto s = (req_state*)lua_touserdata(L, lua_upvalueindex(1));
+
+  if (!lua_isstring(L, -1)) {
+    ldout(s->cct, 1) << "Lua ERROR: field name not provided when accessing S3Request" << dendl;
+    std::cout  << "Lua ERROR: field name not provided when accessing S3Request" << std::endl;
+    return 0;
+  }
+
+  const char* index = lua_tostring(L, -1);
+
+  if (strcmp(index, "op") == 0) {
+    lua_pushinteger(L, s->op);
+  } else if (strcmp(index, "op_type") == 0) {
+    lua_pushinteger(L, s->op_type);
+  } else if (strcmp(index, "content_started") == 0) {
+    lua_pushinteger(L, s->content_started);
+  } else if (strcmp(index, "format") == 0) {
+    lua_pushinteger(L, s->format);
+  } else if (strcmp(index, "decoded_uri") == 0) {
+    lua_pushstring(L, s->decoded_uri.c_str());
+  } else if (strcmp(index, "relative_uri") == 0) {
+    lua_pushstring(L, s->relative_uri.c_str());
+  } else if (strcmp(index, "content_length") == 0) {
+    lua_pushinteger(L, s->content_length);
+  } else if (strcmp(index, "account_name") == 0) {
+    lua_pushstring(L, s->account_name.c_str());
+  } else if (strcmp(index, "bucket_tenant") == 0) {
+    lua_pushstring(L, s->bucket_tenant.c_str());
+  } else if (strcmp(index, "bucket_name") == 0) {
+    lua_pushstring(L, s->bucket_name.c_str());
+  } else {
+  // TODO: rest of the fields
+    ldout(s->cct, 1) << "Lua ERROR: unknown field name " << index << " provided when accessing S3Request" << dendl;
+    std::cout << "Lua ERROR: unknown field name " << index << " provided when accessing S3Request" << std::endl;
+    return 0;
+  }
+  return 1;
+}
+
+int execute(req_state* s, const std::string& script) {
+  auto L = luaL_newstate();
+  lua_state_guard lguard(L);
+
+  luaL_openlibs(L);
+
+  // log action
+  constexpr auto NUMBER_OF_UPVALS = 1;
+  lua_pushlightuserdata(L, s->cct);
+  lua_pushcclosure(L, RGWLog, NUMBER_OF_UPVALS);
+  lua_setglobal(L, "RGWLog");
```

Comment:
is this the debug log or the ops log?

Reply:
this is the debug log, will add the ops log as well

Reply:
@mattbenjamin added ops log